src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs

```diff
@@ -62,140 +55,196 @@ public class HostFilteringMiddleware
+            {
+                MiddlewareHostFilteringOptions options;
+                lock (_syncObject)
+                {
+                    options = _getChangeRequestObject();
+                    _getChangeRequestObject = null;
+                }
+                _middlewareConfiguration = ConfigureMiddleware(options);
+            }
+
+            if (_middlewareConfiguration?.AllowAnyNonEmptyHost != true && (_middlewareConfiguration?.AllowedHosts is null || _middlewareConfiguration.AllowedHosts?.Count == 0))
+            {
+                throw new InvalidOperationException("No allowed hosts were configured.");
+            }
+
+            return _middlewareConfiguration;
+        }
+
+        private MiddlewareConfiguration ConfigureMiddleware(MiddlewareHostFilteringOptions options)
+        {
+            var allowedHosts = new List<StringSegment>();
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                _logger.MiddlewareConfigurationStarted(options.ToString());
+            }
+
+            if (
+                options.AllowedHosts?.Count > 0
+                && !TryProcessHosts(options.AllowedHosts, allowedHosts)
+            )
+            {
+                _logger.WildcardDetected();
+                return new(true, options.AllowEmptyHosts, options.IncludeFailureMessage);
+            }
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                _logger.AllowedHosts(string.Join("; ", allowedHosts));
+            }
+
+            return new(false, options.AllowEmptyHosts, options.IncludeFailureMessage, allowedHosts.AsReadOnly());
+        }
+
+        // returns false if any wildcards were found
+        private static bool TryProcessHosts(IEnumerable<string> incoming, List<StringSegment> results)
+        {
+            foreach (var entry in incoming)
+            {
+                // Punycode. Http.Sys requires you to register Unicode hosts, but the headers contain punycode.
+                var host = new HostString(entry).ToUriComponent();
+
+                if (IsTopLevelWildcard(host))
+                {
+                    // Disable filtering
+                    return false;
+                }
+
+                if (!results.Contains(host, StringSegmentComparer.OrdinalIgnoreCase))
+                {
+                    results.Add(host);
+                }
+            }
+            return true;
+        }
+
+        private static bool IsTopLevelWildcard(string host)
+        {
+            return (
+                string.Equals("*", host, StringComparison.Ordinal) // HttpSys wildcard
+                || string.Equals("[::]", host, StringComparison.Ordinal) // Kestrel wildcard, IPv6 Any
+                || string.Equals("0.0.0.0", host, StringComparison.Ordinal)
+            ); // IPv4 Any
+        }
+
+        private sealed record MiddlewareHostFilteringOptions(ReadOnlyCollection<string> AllowedHosts, bool AllowEmptyHosts, bool IncludeFailureMessage)
```

Comment:
I'm not sure why you need this type either? HostFilteringOptions returned from IOptionsMonitor should be stable.

Reply:
HostFilteringOptions provided from IOptionsMonitor Onchange method is a reference type and it could be changed in any other event handlers, so we should avoid such situations, to achieve this we can create a copy of HostFilteringOptions or use our internal type during evaluation.

Also, the new type could be useful for checking if the new values will be different from the previous ones, and then we can skip the configuration re-evaluation. But it is a good candidate for the future and is not covered by this bug fix.

Reply:
Maybe in theory, but this is the only place this type is consumed so such defense is not needed. We'd rather avoid the extra allocations and work. We'll address it if it becomes an issue in the future.

Reply:
I understand your point as it is really an edge case. But why are you afraid of simple allocation? If users will not update configs often it will be only single allocation. But two quick update can cause various strange behaviors which will be difficult to debug. I can see quite clear benefits of doing it in atomic way and code is more stable.
What do you think?

Reply:
Keep It Simple. Don't solve theoretical problems, only real ones. Unnecessary code makes components harder to understand and maintain.

Reply:
Ok. I will change it as you want